jobrunner/run.py

```diff
@@ -37,42 +38,47 @@
 log = logging.getLogger(__name__)
 
 
+class InvalidTransition(Exception):
+    pass
+
+
 def main(exit_callback=lambda _: False):
     log.info("jobrunner.run loop started")
+
     if config.EXECUTION_API:
         log.info("using new EXECUTION_API")
+        api = job_executor.get_job_api()
+
     while True:
-        active_jobs = handle_jobs()
+        log.debug("Querying database for active jobs")
+        active_jobs = find_where(Job, state__in=[State.PENDING, State.RUNNING])
+        log.debug("Done query")
+        # Randomising the job order is a crude but effective way to ensure that a
+        # single large job request doesn't hog all the workers. We make this
+        # optional as, when running locally, having jobs run in a predictable order
+        # is preferable
+        if config.RANDOMISE_JOB_ORDER:
+            random.shuffle(active_jobs)
+
+        if config.EXECUTION_API:
+            handle_active_jobs_api(active_jobs, api)
+        else:
+            handle_jobs(active_jobs)
+
         if exit_callback(active_jobs):
             break
         time.sleep(config.JOB_LOOP_INTERVAL)
 
 
-def handle_jobs():
-    log.debug("Querying database for active jobs")
-    active_jobs = find_where(Job, state__in=[State.PENDING, State.RUNNING])
-    log.debug("Done query")
-    # Randomising the job order is a crude but effective way to ensure that a
-    # single large job request doesn't hog all the workers. We make this
-    # optional as, when running locally, having jobs run in a predictable order
-    # is preferable
-    if config.RANDOMISE_JOB_ORDER:
-        random.shuffle(active_jobs)
+def handle_jobs(active_jobs):
     for job in active_jobs:
         # `set_log_context` ensures that all log messages triggered anywhere
         # further down the stack will have `job` set on them
         with set_log_context(job=job):
```

Comment:
I think that the loop and the setting of log context are now (still?) identical between API and non-API implementations. Could we pull them up to `main` to remove duplication and reduce the chance of them diverging accidentally?

Reply:
I thought about this, but we use handle_jobs to do one loop through the jobs in tests:

https://github.com/opensafely-core/job-runner/blob/split-run-api/tests/test_integration.py#L71

If we hoist the loop up to `main()`, then the integration test would need to re-implement the loop logic, so another place for divergence.

Not sure which is the lesser evil, tbh!

Reply:
Oh -- if you want it for the tests then you could hoist to a common `handle_jobs()` that does the looping and is called from `main()` and the tests.

Reply:
That's how it was originally :D

Reply:
Have reverted to the original structure

Reply:
You can call `main()` with `exit_callback=lambda _: True` and it will exit after the first iteration -- just to throw another evil into the mix